File: December2022/Java/DetermineIfTwoStringsAreClose.java
// @saorav21994

class Solution {
    public boolean closeStrings(String word1, String word2) {
        
        /*  
            TRUE condition ::
            1. Both words should contain the exact same different characters.
            2. The count of frequency of diff. character in both words should be same.
        */

        int l1 = word1.length();
        int l2 = word2.length();
        
        if (l1 != l2) return false;
        
        int [] f1 = new int[26];
        int [] f2 = new int[26];
        
        for (int i = 0; i < l1; i++) {
            f1[(int)word1.charAt(i)-97] += 1;
            f2[(int)word2.charAt(i)-97] += 1;
        }
        
        for (int i = 0; i < 26; i++) {
            if ((f1[i] > 0 && f2[i] == 0) || (f2[i] > 0 && f1[i] == 0)) return false;
        }
        
        Arrays.sort(f1);
        Arrays.sort(f2);
        
        for (int i = 0; i < 26; i++) {
            if (f1[i] != f2[i]) return false;
        }
        return true;

    }
}

File: December2022/C++/determine-if-two-strings-are-close.cpp
class Solution {
public:
    bool closeStrings(string word1, string word2) {
        
        int n = word1.size();
        
        if(n!=word2.size()){
            return false;
        }
        
       vector<int>freq1(26,0);
        vector<int>freq2(26,0);
        map<int,int>freqWord1;
        map<int,int>freqWord2;
        
        // word1 ->   a->  3   b-> 2 c->5
        // word2 ->  a-> 5  b->3  c-> 1
        
        
        for(int i=0; i<n;i++){
            freq1[word1[i]-'a']++;
        }
        
          for(int i=0; i<n;i++){
            freq2[word2[i]-'a']++;
        }
        
        
        for(int i=0;i<26;i++){
            if(freq1[i]>0 && freq2[i] == 0){
                return false;
            }
             if(freq2[i]>0 && freq1[i] == 0){
                return false;
            }
            freqWord1[ freq1[i]]++;
        
            freqWord2[ freq2[i]]++;
        }
        
        for(auto num:freqWord1){
    
            if(num.second!= freqWord2[num.first]){
                return false;
            }
        }
        
        
        return true;
    }
};